packages/origin-247-certificate/src/offchain-certificate/certificate.aggregate.ts

```diff
@@ -0,0 +1,188 @@
+    }
+
+    public static fromReadModel(certificate: ICertificateReadModel): CertificateAggregate {
+        return new CertificateAggregate({ cert: certificate } as FromReadModel);
+    }
+
+    public static fromEvents(events: ICertificateEvent[]): CertificateAggregate {
+        return new CertificateAggregate({ events: events } as FromEvents);
+    }
+
+    private issue(event: ICertificateEvent): void {
+        if (this.isIssued()) {
+            throw new CertificateErrors.Issuance.CertificateAlreadyIssued(
+                event.internalCertificateId
+            );
+        }
+        const payload = event.payload as IIssueCommand<unknown>;
+
+        this.certificate = {
+            internalCertificateId: event.internalCertificateId,
+            blockchainCertificateId: null,
+            claims: [],
+            creationBlockHash: '',
+            creationTime: Math.floor(Date.now() / 1000),
+            deviceId: payload.deviceId,
+            generationStartTime: Math.floor(payload.fromTime),
+            generationEndTime: Math.floor(payload.toTime),
+            metadata: payload.metadata as any,
+            owners: {
+                [payload.toAddress]: payload.energyValue
+            },
+            claimers: {}
+        };
+    }
+
+    private transfer(event: CertificateTransferredEvent): void {
+        this.validateTransfer(event);
+
+        const { fromAddress, toAddress, energyValue } = event.payload;
+
+        const fromBalance = BigNumber.from(this.certificate!.owners[fromAddress] ?? '0');
+        const toBalance = BigNumber.from(this.certificate!.owners[toAddress] ?? '0');
+
+        const transferVolume = BigNumber.from(energyValue ?? '0');
+        const newFromBalance = fromBalance.sub(transferVolume);
+        const newToBalance = toBalance.add(transferVolume);
+
+        this.certificate!.owners[fromAddress] = newFromBalance.toString();
+        this.certificate!.owners[toAddress] = newToBalance.toString();
+    }
+
+    private claim(event: CertificateClaimedEvent): void {
+        this.validateClaim(event);
+
+        const { forAddress, energyValue, claimData } = event.payload;
+        // claims happen for same address
+        const transferVolume = BigNumber.from(energyValue ?? '0');
+        const unclaimedBalance = BigNumber.from(this.certificate!.owners[forAddress]);
+        const claimedBalance = BigNumber.from(
+            this.certificate!.claimers![forAddress] ? this.certificate!.claimers![forAddress] : '0'
+        );
+
+        const newUnclaimedBalance = unclaimedBalance.sub(transferVolume);
+        const newClaimedBalance = claimedBalance.add(transferVolume);
+
+        this.certificate!.owners[forAddress] = newUnclaimedBalance.toString();
+        this.certificate!.claimers![forAddress] = newClaimedBalance.toString();
+        this.certificate!.claims.push({
+            id: event.internalCertificateId,
+            from: forAddress,
+            to: forAddress,
+            topic: '',
```

Comment:
should this be empty?

Reply:
It's for smart contract, we can leave this empty.